### PR TITLE
refactor(build): no force transpile optional chaining

### PR DESCRIPTION
**Pull request by SukkaW**
- [x] Make sure the linting passes by running `yarn lint`

Ref:

- https://github.com/webpack/webpack/issues/10227
- https://github.com/vercel/next.js/issues/17273
- https://github.com/vercel/next.js/pull/17429
- https://github.com/vercel/next.js/pull/33995
- https://github.com/webpack/webpack/issues/12960
- https://github.com/vercel/next.js/issues/33915
- https://github.com/vercel/next.js/pull/33995
- https://github.com/vercel/next.js/pull/35867

Previously, webpack has some serious issues processing optional chaining (https://github.com/webpack/webpack/issues/10227 which is fixed in webpack 5, and https://github.com/webpack/webpack/issues/12960 which is fixed in webpack 5.71.0). The issues are also being reported to Next.js (https://github.com/vercel/next.js/issues/17273 which is fixed by https://github.com/vercel/next.js/pull/17429, and https://github.com/vercel/next.js/issues/33915 which is fixed by https://github.com/vercel/next.js/pull/33995).

With Next.js has deprecated Webpack 4 since this major version and has updated Webpack to 5.71.0 (https://github.com/vercel/next.js/pull/35867), it is now safe to remove force transpilation of optional chaining and nullish operator.

### CODE REVIEW

File: test/unit/next-babel-loader-prod.test.ts [hunk 1]
```diff
@@ -323,75 +323,5 @@ describe('next-babel-loader', () => {
         `var __jsx = React.createElement;import "core-js";import { bar } from "a";import baz from "b";import * as React from "react";import { yeet } from "c";import baz3, { cats } from "d";import { c, d } from "e";import { e as ee } from "f";export var __N_SSG = true;export default function () {  return __jsx("div", {    __self: this,    __source: {      fileName: _jsxFileName,      lineNumber: 1,      columnNumber: 326    }  }, cats + bar());}`
       )
     })
-
-    it('should support optional chaining for JS file', async () => {
```
Comment (SukkaW) [inline test/unit/next-babel-loader-prod.test.ts:327]: It is safe to remove those unit test cases. Next.js has an another integration test case `test/integration/optional-chaining-nullish-coalescing` to make sure a Next.js application with optional chaining and nullish operator will work.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| buildDuration | 15.4s | 15.4s | -95ms |
| buildDurationCached | 6.1s | 6.1s | ⚠️ +31ms |
| nodeModulesSize | 477 MB | 477 MB | -1.17 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.073 | 3.063 | -0.01 |
| / avg req/sec | 813.63 | 816.31 | +2.68 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.208 | 1.172 | -0.04 |
| /error-in-render avg req/sec | 2070.31 | 2132.52 | +62.21 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.7 kB | 71.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.05 kB | 3.05 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.64 kB | 5.64 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.32 kB | 2.32 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| buildDuration | 18.3s | 18s | -279ms |
| buildDurationCached | 6.1s | 6.1s | ⚠️ +38ms |
| nodeModulesSize | 477 MB | 477 MB | -1.17 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.082 | 3.057 | -0.02 |
| / avg req/sec | 811.18 | 817.79 | +6.61 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.207 | 1.184 | -0.02 |
| /error-in-render avg req/sec | 2071.24 | 2111.94 | +40.7 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.3 kB | 28.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.2 kB | 72.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.03 kB | 3.03 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.7 kB | 5.7 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: f08683bb6cde1a922c36cf3bacd42e6b6ec7eeec

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| buildDuration | 15.5s | 15.8s | ⚠️ +316ms |
| buildDurationCached | 6.1s | 6s | -97ms |
| nodeModulesSize | 478 MB | 478 MB | -1.17 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.183 | 3.193 | ⚠️ +0.01 |
| / avg req/sec | 785.34 | 782.85 | ⚠️ -2.49 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.351 | 1.356 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1851.12 | 1843.5 | ⚠️ -7.62 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.7 kB | 71.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.05 kB | 3.05 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.68 kB | 5.68 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.32 kB | 2.32 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.9 kB | 15.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| buildDuration | 18.4s | 18.3s | -131ms |
| buildDurationCached | 6s | 6.1s | ⚠️ +40ms |
| nodeModulesSize | 478 MB | 478 MB | -1.17 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.2 | 3.166 | -0.03 |
| / avg req/sec | 781.23 | 789.74 | +8.51 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.346 | 1.354 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1856.89 | 1846.74 | ⚠️ -10.15 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.3 kB | 28.3 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.2 kB | 72.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.03 kB | 3.03 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.74 kB | 5.74 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | SukkaW/next.js disable-nullish-optional-chain-transform | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: f356d0327998e275dbc2f8ea99fbb0c63127882b

**Review by timneutkens — COMMENTED** [review]
Is optional chaining still downleveled when the targeted level doesn't support it? If that's the case these changes look fine.

**Comment by SukkaW** [comment]
> Is optional chaining still downleveled when the targeted level doesn't support it? If that's the case these changes look fine.

@timneutkens Check `test/integration/optional-chaining-nullish-coalescing`. If that integrated test case passed, then it is fine.

File: test/unit/next-babel-loader-prod.test.ts (hunk 1 above)
Comment (ijjk) [reply test/unit/next-babel-loader-prod.test.ts:327]: Let's keep these unit tests just for a sanity check

**Review by ijjk — COMMENTED** [review]
This change looks alright now as the test added in https://github.com/vercel/next.js/pull/33995 is passing against Node.js v16 and v17 which is what the affected versions were. I think it would be good to keep the unit tests though

**Comment by SukkaW** [comment]
> This change looks alright now as the test added in #33995 is passing against Node.js v16 and v17 which is what the affected versions were. I think it would be good to keep the unit tests though

The Next.js 12 requires minimum Node.js version of 12.22.0, which doesn't support optional chaining syntax. And the CI run tests on Node.js 16 & 17 which both have optional chaining syntax support.

IMHO the behavior should be tested by swc and babel (whether they can identify and apply transpilation), not by Next.js.

**Review by ijjk — APPROVED** [review]

**Pull request merged**